packages/evm/src/hybridStack.ts

```diff
@@ -0,0 +1,314 @@
+ */
+
+type StackElement = [bigint | undefined, Uint8Array | undefined]
+type Store = StackElement[]
+
+export class HybridStack implements EVMStack {
+  // This array is initialized as an empty array. Once values are pushed, the array size will never decrease.
+  private _store: Store
+  private _maxHeight: number
+
+  private _len: number = 0
+
+  constructor(maxHeight?: number) {
+    // It is possible to initialize the array with `maxHeight` items. However,
+    // this makes the constructor 10x slower and there do not seem to be any observable performance gains
+    this._store = []
+    this._maxHeight = maxHeight ?? 1024
+  }
+
+  get length() {
+    return this._len
+  }
+
+  /**
+   * Push an item to the stack.
+   * @deprecated use the type-named version of the method
+   * @param value
+   */
+  push(value: bigint) {
+    this.pushBigInt(value)
+  }
+
+  /**
+   * Push an item to the stack.
+   * @param value
+   */
+  pushBigInt(value: bigint) {
+    if (this._len >= this._maxHeight) {
+      throw new EvmError(ERROR.STACK_OVERFLOW)
+    }
+
+    // Read current length, set `_store` to value, and then increase the length
+    this._store[this._len++] = [value, undefined]
+  }
+
+  /**
+   * Push an item to the stack.
+   * @param value
+   */
+  pushBytes(value: Uint8Array) {
+    if (this._len >= this._maxHeight) {
+      throw new EvmError(ERROR.STACK_OVERFLOW)
+    }
+
+    // Read current length, set `_store` to value, and then increase the length
+    this._store[this._len++] = [undefined, value]
+  }
+
+  /**
+   * Pop an item from the stack.
+   * @deprecated use the type-named version of the method
+   * @param value
+   */
+  pop(): bigint {
+    return this.popBigInt()
+  }
+
+  /**
+   * Pop an item from the stack.
+   * @param value
+   */
+  popBigInt(): bigint {
+    if (this._len < 1) {
+      throw new EvmError(ERROR.STACK_UNDERFLOW)
+    }
+
+    // Length is checked above, so pop shouldn't return undefined
+    // First decrease current length, then read the item and return it
+    // Note: this does thus not delete the item from the internal array
+    // However, the length is decreased, so it is not accessible to external observors
+    const elem = this._store[--this._len]
+    if (elem[0] !== undefined) {
+      return elem[0]
+    }
+    return bytesToBigInt(elem[1]!)
+  }
+
+  /**
+   * Pop an item from the stack.
+   * @param value
+   */
+  popBytes(): Uint8Array {
+    if (this._len < 1) {
+      throw new EvmError(ERROR.STACK_UNDERFLOW)
+    }
+
+    // Length is checked above, so pop shouldn't return undefined
+    // First decrease current length, then read the item and return it
+    // Note: this does thus not delete the item from the internal array
+    // However, the length is decreased, so it is not accessible to external observors
+    const elem = this._store[--this._len]
+    if (elem[1] !== undefined) {
+      return elem[1]
+    }
+    return bigIntToBytes(elem[0]!)
```

Comment:
Same question as above.  Shouldn't we just return `elem[1]` here and leave it up to the consumer to know that it's bytes vs bigint?